pkg/operator/controller/ingress/controller.go

```diff
@@ -1131,6 +1134,18 @@ func (r *reconciler) ensureIngressController(ci *operatorv1.IngressController, d
 
 	SetIngressControllerNLBMetric(ci)
 
+	// If the lbService exists, then update Infra CR's PlatformStatus with the Ingress LB IPs
+	if haveLB {
+		if updated, err := updateInfraIngressLoadBalancerIPs(lbService, infraConfig); err != nil {
+			errs = append(errs, fmt.Errorf("failed to update Infrastructure PlatformStatus: %w", err))
+		} else if updated {
+			if err := r.client.Status().Update(context.TODO(), infraConfig); err != nil {
```

Comment:
This code is a little surprising to me because usually in this operator `updateFoo()` performs the update itself.  Could you move the `r.client.Status().Update()` call to `updateInfraIngressLoadBalancerIPs` itself for consistency with other code?

For completeness, this is the pattern we use in this operator:
<details>

```go
func (r *reconciler) ensureFoo(...) (bool, *foov1.Foo, error) {
	haveFoo, current, err := r.currentFoo()
	if err != nil {
		return false, nil, err
	}

	wantFoo, desired, err := desiredFoo()
	if err != nil {
		return false, nil, err
	}

	switch {
	case !wantFoo && !haveFoo:
		return false, nil, nil
	case !wantFoo && haveFoo:
		// Delete current.
	case wantFoo && !haveFoo:
		// Create desired.
	case wantFoo && haveFoo:
		// Update current to match desired.
		if updated, err := r.updateFoo(current, desired); err != nil {
			return true, current, err
		} else if updated {
			return r.currentFoo()
		}
	}

	return true, current, nil
}

func desiredFoo() (bool, *foov1.Foo, error) {
	foo := &corev1.Foo{
		// ...
	}

	return true, foo, nil
}

func (r *reconciler) updateFoo(current, desired *corev1.Foo) (bool, error) {
	changed, updated := fooChanged(current, desired)
	if !changed {
		return false, nil
	}

	// Diff before updating because the client may mutate the object.
	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
	if err := r.client.Update(context.TODO(), updated); err != nil {
		return false, err
	}
	log.Info("updated foo", "namespace", updated.Namespace, "name", updated.Name, "diff", diff)

	return true, nil
}

func fooChanged(current, expected *corev1.Foo) (bool, *corev1.Foo) {
	changed := false
	fooCmpOpts := []cmp.Option{
		// ...
	}
	if !cmp.Equal(current.Spec, expected.Spec, fooCmpOpts...) {
		changed = true
	}

	if !changed {
		return false, nil
	}

	updated := current.DeepCopy()
	// ...

	return true, updated
}
```
</details>

We usually have unit tests for `fooChanged` (and sometimes for `desiredFoo` as well).

Reply:
I followed the pattern seen in https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/controller.go#L1119-L1130

Reply:
`syncIngressControllerStatus` actually does an update using `r.client.Status().Update(context.TODO(), updated)`, and then `ensureIngressController` does a `r.client.Get(...)` so it has the newly updated object.  

My point is that with `syncIngressControllerStatus` (in your example) and `updateFoo` (in my example), the name suggests that the function updates the API object in etcd, and the function does exactly that.  However, this new `updateInfraIngressLoadBalancerIPs` returns a pointer to a `configv1.Infrastructure` value without actually updating object in etcd.  

So I'd prefer this new `updateInfraIngressLoadBalancerIPs` function either call `r.client.Status().Update(context.TODO(), infraConfig)` itself, or that `updateInfraIngressLoadBalancerIPs` be named something different, such as `computeUpdatedInfraFromService`.